Accumulo 4510

mikewalch opened the pull request:
To make this review easier, I pushed the commit ac5b271caec45083751efad7298852969bec387e to master that copied the code from Accumulo.  The commit has some changes to package names and POM file so feel free to review.  

This PR contains code that improves how random walk tests are run.  Users can now run them locally or in YARN.

joshelser commented on README.md:
```diff
@@ -1 +1,50 @@
+`accumulo-testing.props` to the number of walkers that you want started and run the command
+below.
+
+        ./bin/accumulo-testing rw-cluster Image.xml
```
I'd suggest `rw-yarn` or `rw-twill` instead of `rw-cluster` for some future proofing.

mikewalch replied:
`rw-yarn` sounds good to me.  I will change.

joshelser commented on README.md:
```diff
@@ -1 +1,50 @@
+## Random walk test
+
+The random walk test generates client behavior on an Apache Accumulo instance by randomly walking a
+graph of client operations. The test is configured by the `accumulo-testing.props` file and requires
```
Define here what the minimum set of properties are? Also, would prefer the standard "properties" suffix, but that's just me being picky.

mikewalch replied:
Yes min set of properties is needed and I will change suffix to standard.

joshelser commented on bin/accumulo-testing:
```diff
@@ -0,0 +1,89 @@
+  . $AT_HOME/conf/accumulo-testing-env.sh.example
+fi
+
+export AT_PROPS=$AT_HOME/conf/accumulo-testing.props
```
shellcheck: quotes needed.

mikewalch replied:
will fix

joshelser commented on bin/accumulo-testing:
```diff
@@ -0,0 +1,89 @@
+    exit 1
+  fi
+  build_shade_jar
+  mvn compile -Ptest-runner -Dhadoop.version=$HADOOP_VERSION -Dexec.args="-n AccumuloRandomWalkTest -j $AT_SHADED_JAR -m $randomwalk_main -t $AT_PROPS -l $log4j_config -a ./accumulo-testing.props $2"
```
compile? Not package?

joshelser replied:
Hm. @ctubbsii knows off the top of his head if this would cause any problems? Maybe `exec:java` can use the compiled class files (and doesn't require a jar).

mikewalch replied:
See yarn/pom.xml for the exec-maven-plugin config.  I cannot just use `mvn exec:java` on the command line as the class that I want to execute is in a module.  The execution is being attached to the compile phase as a jar isn't required.

ctubbsii replied:
The `build_shade_jar` function does a `mvn package`, and the `mvn compile` here is just to trigger the `exec:java` from the compiled code. I'm sure there's a more efficient way to do this, because this seems a bit redundant, but I can't see any immediate problems with it.

joshelser replied:
Thanks for the explanation, Mike and Christopher. +1

joshelser commented on conf/accumulo-testing-env.sh.example:
```diff
@@ -0,0 +1,25 @@
+# ========
+# Set the Accumulo version that should be included in the shaded jar
+export ACCUMULO_VERSION=`accumulo version`
+export HADOOP_VERSION=`hadoop version | grep Hadoop | cut -d ' ' -f 2`
```
I think `hadoop version | head -n1 | awk '{print $2}'` is a little more clear here.

joshelser commented on twill/src/main/java/org/apache/accumulo/testing/twill/TestRunner.java:
```diff
@@ -0,0 +1,166 @@
+      controller.awaitTerminated();
+      done.set(true);
+    } catch (ExecutionException e) {
+      LOG.error("Error", e);
```
Wouldn't we want to re-throw the exception so that we get a non-zero exit code?

joshelser commented on twill/src/main/java/org/apache/accumulo/testing/twill/TestRunner.java:
```diff
@@ -0,0 +1,166 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class TestRunner {
```
More descriptive class name, please.

joshelser commented on twill/pom.xml:
```diff
@@ -0,0 +1,118 @@
+      <dependency>
+        <groupId>com.google.guava</groupId>
+        <artifactId>guava</artifactId>
+        <version>13.0.1</version>
```
Why not have this defined in `accumulo-testing`?

mikewalch replied:
The guava version in the main pom is set to `14.0.1` which coincides with Accumulo.  Twill has a hard requirement on `13.0.1`.  Things break even if `14.0.1` is used.  I tried this and the version can be moved down to where guava is listed in dependencies.  I will also add comment.

joshelser commented on bin/accumulo-testing:
```diff
@@ -0,0 +1,89 @@
+    cd $AT_HOME
+
+    : ${ACCUMULO_VERSION?"ACCUMULO_VERSION must be set in bash env or conf/accumulo-testing-env.sh"}
+    mvn clean package -Pcreate-shade-jar -DskipTests -Daccumulo.version=$ACCUMULO_VERSION -Dhadoop.version=$HADOOP_VERSION
```
Tricky ;)

We should be making sure that the jar which is created is properly licensed (that is un-skirtable). Have you taken a look to see if it looks "sane"?

mikewalch replied:
I took a look at and it looks sane to me.

joshelser commented on bin/accumulo-testing:
```diff
@@ -0,0 +1,89 @@
+    echo "Building $AT_SHADED_JAR"
+    cd $AT_HOME
+
+    : ${ACCUMULO_VERSION?"ACCUMULO_VERSION must be set in bash env or conf/accumulo-testing-env.sh"}
```
I actually have no clue what this colon is doing here. Is this a typo or just some weird bash-ism I'm not familiar with?

joshelser replied:
OK. Definitely don't change the bash if it works just because I don't recognize it. If you could slap in a comment as to what it's doing, that's more than enough for me :)

mikewalch replied:
no worries.  i changed it.  the check is higher up in the code now after accumulo-testing-env.sh is sourced.

joshelser commented:
Thanks, Mike! I'll try to take another look, but feel free to push now if you'd like. I trust your judgement.

keith-turner commented on conf/accumulo-testing-env.sh.example:
```diff
@@ -0,0 +1,25 @@
+
+# Accumulo
+# ========
+# Set the Accumulo version that should be included in the shaded jar
```
would be nice to make the version of zookeeper included in the shaded jar configurable also

mikewalch replied:
this added in 45b9dbfbbca

keith-turner commented on README.md:
```diff
@@ -1 +1,54 @@
+
+This command will create an application in YARN and exit when test is completed. While its
+running, you can view logs for each random walker using the YARN resource manager. The YARN
+application can be killed at any time using `ctrl-c` or via the resource manager.
```
If you kill the application driving the test using `ctrl-c`, will that cause things in yarn to die?

mikewalch replied:
yes it will.

keith-turner replied:
So can I ssh to a cluster, start random walk on N node, and then log off?  Maybe I need to run the application that stated the job with nohup.

mikewalch commented:
I think that this pull request is ready to merged. Before this is merged, I want to mention two more commits that were added to PR in case any one wants to review:

* 45b9dbfbbcaf - I fixed a bug that was preventing RW tests from launching MapReduce jobs.  I also removed test code that depended on Accumulo being installed locally.
* 36f31026561de99973e - I changed the behaviour of the `accumulo-testing rw-yarn` command to exit after all containers of the YARN application have been created but before test has ended.
